presentation/backend/src/presentation-backend/RulesetEmbedder.ts

```diff
@@ -164,32 +191,47 @@ export class RulesetEmbedder {
     // attempt to update ruleset with same ID and version
     const exactMatch = rulesetsWithSameId.find((curr) => curr.normalizedVersion === rulesetVersion);
     if (exactMatch !== undefined) {
-      return this.updateRuleset(exactMatch.id, ruleset);
+      return this.updateRuleset(exactMatch.id, ruleset, normalizedOptions.onElementUpdate);
     }
 
     // no exact match found - insert a new ruleset element
-    const model = this.getOrCreateRulesetModel();
+    const model = this.getOrCreateRulesetModel(normalizedOptions.onElementInsert, normalizedOptions.onModelInsert);
     const rulesetCode = RulesetElements.Ruleset.createRulesetCode(this._imodel, model.id, ruleset);
-    return this.insertNewRuleset(ruleset, model, rulesetCode);
+    return this.insertNewRuleset(ruleset, model, rulesetCode, normalizedOptions.onElementInsert);
   }
 
-  private updateRuleset(elementId: Id64String, ruleset: Ruleset) {
+  private updateRuleset(elementId: Id64String, ruleset: Ruleset, onElementUpdate?: OnElementUpdate) {
     const existingRulesetElement = this._imodel.elements.tryGetElement<DefinitionElement>(elementId);
     assert(existingRulesetElement !== undefined);
     existingRulesetElement.jsonProperties.jsonProperties = ruleset;
-    existingRulesetElement.update();
+
+    onElementUpdate?.onBeforeElementUpdate();
+    try {
+      existingRulesetElement.update();
+    } finally {
+      onElementUpdate?.onAfterElementUpdate();
+    }
+
     this._imodel.saveChanges();
     return existingRulesetElement.id;
   }
 
-  private insertNewRuleset(ruleset: Ruleset, model: Model, rulesetCode: Code): Id64String {
+  private insertNewRuleset(ruleset: Ruleset, model: Model, rulesetCode: Code, onElementInsert?: OnElementInsert): Id64String {
     const props: DefinitionElementProps = {
       model: model.id,
       code: rulesetCode,
       classFullName: RulesetElements.Ruleset.classFullName,
       jsonProperties: { jsonProperties: ruleset },
     };
-    const id = this._imodel.elements.insertElement(props);
+
+    let id;
+    onElementInsert?.onBeforeElementInsert();
+    try {
+      id = this._imodel.elements.insertElement(props);
+    } finally {
+      onElementInsert?.onAfterElementInsert();
+    }

```

Comment:
Instead of repeating a similar piece of code in multiple places, could we instead extract it to methods `insertElement<TProps>(props: TProps, callbacks?: InsertCallbacks<TProps>): Id64String` and `updateElement<TElement>(element: TElement, callbacks?: UpdateCallbacks<TElement>)`?

Reply:
Updated